### PR TITLE
Add support for conditional breakpoints

### PR DESCRIPTION
**Pull request by isc-bsaviano**
This PR adds support for [conditional breakpoints.](https://code.visualstudio.com/docs/editor/debugging#_conditional-breakpoints) Note that due to a limitation in the Atelier debugger implementation, we can only support hit conditions that are integers instead of more complex expressions.

### CODE REVIEW

**Review by gjsjohnmurray — COMMENTED** [review]
I'm going to recuse myself from reviewing this because of a conflict of interests with our own debugger.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray That's reasonable.

@daimor Can you please give this a look before our meeting tomorrow?

File: src/debug/debugSession.ts [hunk 1]
```diff
@@ -292,8 +323,17 @@ export class ObjectScriptDebugSession extends LoggingDebugSession {
       await Promise.all(
         xdebugBreakpoints.map(async (breakpoint, index) => {
           try {
-            await this._connection.sendBreakpointSetCommand(breakpoint);
-            vscodeBreakpoints[index] = { verified: true, line: breakpoint.line };
+            if (breakpoint.hitCount && !/^\d+$/.test(breakpoint.hitCount)) {
```
Comment (daimor) [inline src/debug/debugSession.ts:326]: If `hitCount` allows only integer, why has a `string` type?

File: src/debug/xdebugConnection.ts [hunk 2]
```diff
@@ -206,7 +209,7 @@ export class ClassLineBreakpoint extends LineBreakpoint {
       this.line = parseInt(breakpointNode.getAttribute("lineno"), 10);
       this.fileUri = breakpointNode.getAttribute("filename");
     } else {
-      super(rest[0], rest[1]);
+      super(rest[0], rest[1], rest[4]);
```
Comment (daimor) [inline src/debug/xdebugConnection.ts:212]: why not just
`parseInt(rest[4], 10)`
parseInt will return `NaN` for anything else

File: src/debug/xdebugConnection.ts [hunk 3]
```diff
@@ -157,6 +159,7 @@ export abstract class Breakpoint {
       this.state = breakpointNode.getAttribute("state") as BreakpointState;
     } else {
       this.type = rest[0];
+      this.hitCount = rest[1];
```
Comment (daimor) [inline src/debug/xdebugConnection.ts:162]:
```suggestion
      this.hitCount = rest[1] ? parseInt(rest[1], 10) : null;
```

File: src/debug/xdebugConnection.ts [hunk 4]
```diff
@@ -143,10 +143,12 @@ export abstract class Breakpoint {
   public state: BreakpointState;
   /** The connection this breakpoint is set on */
   public connection: Connection;
+  /** A numeric value used to determine if the breakpoint should pause execution or be skipped. */
+  public hitCount?: string;
```
Comment (daimor) [inline src/debug/xdebugConnection.ts:147]:
```suggestion
  public hitCount?: number;
```

File: src/debug/xdebugConnection.ts [hunk 5]
```diff
@@ -143,10 +143,12 @@ export abstract class Breakpoint {
...
   /** Constructs a breakpoint object from an XML node from a XDebug response */
   public constructor(breakpointNode: Element, connection: Connection);
   /** To create a new breakpoint in derived classes */
-  public constructor(type: BreakpointType);
+  public constructor(type: BreakpointType, hitCount?: string);
```
Comment (daimor) [inline src/debug/xdebugConnection.ts:151]:
```suggestion
  public constructor(type: BreakpointType, hitCount?: number);
```

**Review by daimor — CHANGES_REQUESTED** [review]

**Comment by isc-bsaviano** [comment]
@daimor I renamed `hitCount` to `hitCondition` and updated its description to better represent what I'm using it for.

File: src/debug/debugSession.ts [hunk 6]
```diff
@@ -323,12 +323,12 @@ export class ObjectScriptDebugSession extends LoggingDebugSession {
       await Promise.all(
         xdebugBreakpoints.map(async (breakpoint, index) => {
           try {
-            if (breakpoint.hitCount && !/^\d+$/.test(breakpoint.hitCount)) {
+            if (breakpoint.hitCondition && !/^\d+$/.test(breakpoint.hitCondition)) {
```
Comment (daimor) [inline src/debug/debugSession.ts:326]: you can still use `parseInt` here instead of regex

**Review by daimor — CHANGES_REQUESTED** [review]

**Review by daimor — APPROVED** [review]

**Pull request merged**